Add autocomplete users turn functionality

LukeKuenneke opened the pull request:
Opening the options menu will allow the user to disable during turn.

yairm210 commented on core/src/com/unciv/ui/worldscreen/WorldScreen.kt:
```diff
@@ -208,12 +209,31 @@ class WorldScreen(val viewingCiv:CivilizationInfo) : CameraStageBaseScreen() {
+                viewingCiv.popupAlerts.none() &&
+                viewingCiv.tradeRequests.isEmpty() &&
+                !viewingCiv.shouldOpenTechPicker() &&
+                viewingCiv.getIdleUnits().isEmpty() &&
```
Isn't there a "IsSomethingOpen" function that covers many of these? You could probably use that

LukeKuenneke replied:
I will have to go back and give this a look another time, I definitely want the readability of this function to be better.

yairm210 commented:
In general, looks good!
I assume we'll have to figure out the circumstances of 'when to next turn' with real users. so it's good that it's opt-in ;)

yairm210 commented:
Ignore my comment on the "IsSomethingOpen", that was replaced by the "hasVisibleDialogs" function
Please add the "Autocomplete turn" text to the Android/assets/jsons/translationsByLanguage/template.properties file for translation purposes, and we'll be ready to go!

LukeKuenneke commented:
> Ignore my comment on the "IsSomethingOpen", that was replaced by the "hasVisibleDialogs" function
> Please add the "Autocomplete turn" text to the Android/assets/jsons/translationsByLanguage/template.properties file for translation purposes, and we'll be ready to go!

Done.

LukeKuenneke commented:
> In general, looks good!
> I assume we'll have to figure out the circumstances of 'when to next turn' with real users. so it's good that it's opt-in ;)

I have some checks for idle units, research available, and ability to adopt a policy in the canAutoCompleteTurn function. 

It will be a good thing to get feedback on for sure.

r3versi commented on core/src/com/unciv/ui/worldscreen/WorldScreen.kt:
```diff
@@ -208,12 +209,31 @@ class WorldScreen(val viewingCiv:CivilizationInfo) : CameraStageBaseScreen() {
+                !viewingCiv.shouldOpenTechPicker() &&
+                viewingCiv.getIdleUnits().isEmpty() &&
+                stage.actors.last() !is WorldScreenMenuTable
+    }
```
Instead of `!viewingCiv.policies.canAdoptPolicy()` I think you should check for `!viewingCiv.policies.shouldOpenPolicyPicker` (so that the autoturn can go on if the player knows he can adopt a policy and refuses to).

Instead of `viewingCiv.getIdleUnits().isEmpty()` you should call `viewingCiv.shouldGoToDueUnit()` (there is an option to ignore the idle units).

r3versi commented on core/src/com/unciv/ui/worldscreen/optionstable/WorldScreenOptionsTable.kt:
```diff
@@ -77,6 +77,12 @@ class WorldScreenOptionsTable(val worldScreen:WorldScreen) : PopupTable(worldScr
+        innerTable.addButton(if (settings.autoCompleteTurn) "Yes".tr() else "No".tr()) {
+            settings.autoCompleteTurn = !settings.autoCompleteTurn
+            update()
+        }
```
I think this option should be in the _Gameplay options_ section.

yairm210 replied:
Oh yeah
Good point, missed that =)

Iver88 commented:
Please don't do same bugged version like in Civ5 and Civ6 where player isn't able to stop this feature and it doesn't stop on some important events. Promise me :D

rh-github-2015 commented:
A completely different approach to the usecase might be to have a long-press on the "next turn button that is currently not the next turn button" always trigger next turn.

Is Unciv already using GestureListener? Must check later.

lishaoxia1985 commented:
@LukeKuenneke i don't think it's necessary because the tech  and the building picked should be changed in anytime,  the end of the turn must be up to players😄.

LukeKuenneke commented:
Closing PR.

Undecided UX.

LukeKuenneke commented:
@yairm210 I almost forgot about this, is this functionality still out of scope?

yairm210 commented:
Done by @tuvus including addressing the points raised in this - your work was not in vain, it gave us the direction to go in :)

LukeKuenneke commented:
> Done by @tuvus including addressing the points raised in this - your work was not in vain, it gave us the direction to go in :)

No worries, glad something came of it :)
